README.md
# autoresponse [![NPM Version](https://img.shields.io/npm/v/autoresponse.svg?style=flat)](https://npmjs.org/package/autoresponse)

> A connect middleware for mocking the http request, can used in `edp-webserver` or `webpack-dev-server` mocking

## Install

**Require Node.js 8+**

```shell
npm install autoresponse
```

## Usage

The `autoresponse` mock config, you can specify by a `autoresponse-config.js` config file or passing the mock config params.

```javascript
var autoresponse = require("autoresponse")({
    logLevel: "info", // the level to print log info
    post: true, // mock all post request
    patch: true, // mock all patch request
    get: {
        match: function(reqPathName) {
            // mock all `/xx/xx` path
            return !/\.\w+(\?.*)?$/.test(reqPathName);
        }
    },

    // or using rules options, support request methods setting
    // the rules option priority is higher than the outside `get`/`post`/... request
    // method mock rule
    rules: [
        {
            // by default mock all methods, if method option is not set
            match: "/users/:id", // by default use `users.js` mock file
            method: ["get", "patch"]
        },
        {
            match: function(reqPathName, reqMethod) {
                return true;
            },
            mock: function(reqPathName, reqMethod) {
                return "custom/myMockFile.js";
            },
            method: "post"
        }
    ]
});
```

By default the mock file path is the same as the request url pathname, e.g., the request pathname is `/a/b`, the default mock file path is `<projectRoot>/mock/a/b.js`. If the mock file is not existed, `autoresponse` will auto create the mock file basing the mock file template.

The mock file like this:

```javascript
module.exports = function(path, queryParam, postParam, context) {
    return {
        timeout: 50, // response timeout, unit is millisecond, default is 0
        _timeout: 50, // The same as timeout
        _status: 200, // The response http status code, by default 200
        _header: {}, // The response header
        _data: {}, // The response mock data
        _jsonp: false, // response jsonp
        _callback: "callback" // The jsonp callback param name, default: callback
    };

    // mock data placed in `_data` is not required, the following is also valid
    // return {
    //    timeout: 10,
    //    status: 0,
    //    statusInfo: {}
    // };
};
```

If the same request need to mock different request method, you can also write the mock data, like this:

```javascript
module.exports = {
    // mock the post request
    post: function(path, queryParam, postParam, context) {
        var params = context.params; // the restful path param
        return {
            // ...
        };
    },

    // mock the patch request
    patch: {
        status: 0,
        statusInfo: "patch ok"
    }
};
```

`Autoresponse` supports any file types mocking, you can using `js file`, `json file` or any other custom mock syntax to generate the mock data. For example, you can using `js` to mock `smarty` template without needing `php` programming. If there is none available mock handler, you can also custom it by yourself.

Moreover, `autoresponse` provide some useful [mock helpers](#helper) to help generating mock data.

**Tip:** If you need modify the mock config frequently, the best choice is using `autoresponse-config.js` config file, `autoresponse` support auto reload config file by using `watch: true` option without having to restart the dev server. The more usage information you can see [here](#config-file).

The more detail usage, you can see [examples](https://github.com/wuhy/autoresponse/tree/master/examples).

## Using as a connect middleware

```javascript
var autoresponse = require("autoresponse")({
    logLevel: "info",
    post: true
});
app.use(autoresponse);

var serveStatic = require("serve-static");
app.use(serveStatic("./webroot"));
```

## Using in webpack-dev-server

[webpack-dev-server](https://github.com/webpack/webpack-dev-server) is developed based on `express`, so `autoresponse` can as a middleware served it using `setup` option.

```javascript
var compiler = Webpack(webpackConfig);
var server = new WebpackDevServer(compiler, {
    // install middlewares，if your webpack version is less than 3.0.0, using `setup` option instead of `before` option
    before: function(app) {
        var autoresponse = require("autoresponse");
        app.use(
            autoresponse({
                logLevel: "debug",
                root: projectRootPath, // you can specify the project root path
                post: true, // mock all post request
                patch: true // mock all patch request
            })
        );
    }
});

server.listen(8888, function() {
    console.log("Starting server on port 8888...");
});
```

## Using in edp-webserver

If you use [EDP](https://github.com/ecomfe/edp) solution, you can also use `autoresponse` middleware in [edp-webserver](https://github.com/ecomfe/edp-webserver) to mock the http request.

```javascript
exports.getLocations = function() {
    return [
        {
            location: "/",
            handler: home("index.html")
        },
        {
            location: /\.html\b.*$/,
            handler: [file()]
        },
        // add autoresposne mock handler
        require("autoresponse")("edp", { watch: true, logLevel: "info" }),
        {
            location: /^.*$/,
            handler: [file(), proxyNoneExists()]
        }
    ];
};
```

## Using mock config file <a name="config-file"></a>

Create `autoresponse` middleware:

```javascript
var autoresponse = require("autoresponse")({
    // specify whether need auto reload config file when config file change
    watch: true
});
```

Create `autoresponse-config.js` file in your web document root, the config file content like the following:

```javascript
module.exports = {
    // The response directory to mock, by default is `mock`
    responseDir: "./mock",

    /**
     * configure the `get` request, determine the request path and the file to be mocked.
     * You can also configure the `post` request and `query` params to mock.
     * More information, please refer to examples.
     *
     * @type {boolean|Array}
     */
    get: [
        {
            match: "/b.html",
            mock: "c.html"
        },
        {
            match: "/account/getUserInfo", // also support regex and function
            mock: {
                proxy: "localhost:9090" // use proxy
            }
        },
        {
            // default mock file: <responseDir>/user/profile.js
            // it'will be processed as a node module by builtin js-processor
            match: "/user/profile"
        },
        {
            match: "/data/list",
            mock: "data/list.json"
        },
        {
            match: "/php",
            mock: {
                path: "/data/test.php" // rewrite request path
            }
        },
        {
            match: "/a/b",
            mock: "a/b.php" // mock with php file which is processed by php processor
        },
        "/account/getUserInfo", // specify the match path
        function(reqPath, context) {
            // using function to determine which request to mock
            return {
                match: "a/b"
            };
        }
    ]
};
```

## Using smarty processor

Convert `json` data to `html` or `html segment` using `smarty` template engine.

**prepare：**

-   install php-cgi

    -   [for mac](https://gist.github.com/xiangshouding/9359739)
    -   [for windows](https://gist.github.com/lily-zhangying/9295c5221fa29d429d52)

-   processor configure

```javascript
// add this config to autoresponse
processor: {
    smarty: {
        // specify the initialize configure file, the file path is relative to `responseDir`
        // the file content you can refer below
        initerFile: './initer.php',

        // you also can specify your php-cgi path here, default using `php-cgi`
        php: {bin: '<php-cgi path>'}
    }
}
```

```php
<?php

// ============== initer.php ================

error_reporting(0);
ini_set('error_reporting', E_ALL & ~E_NOTICE);

$project_root = dirname(dirname(__FILE__));

require dirname($project_root) . '/libs/Smarty.class.php';

// initialize the smarty variable
$smarty = new Smarty;

$smarty->force_compile = true;
$smarty->debugging = false;
$smarty->caching = false;

// specify the delimiter chararter
$smarty->left_delimiter = '{';
$smarty->right_delimiter = '}';

// setting the template directory and output directory for compilation
$smarty->setTemplateDir($project_root . '/templates/');
$smarty->setCompileDir($project_root . '/templates_c/');
```

-   write smarty json data using js processor

    -   output html document

        ```javascript
        module.exports = function(path, queryParam, postParam) {
            return {
                // of course, you can specify the delay time with a random value between 0 and 100
                _timeout: "0,100",

                // if you wanna simulate the special status, you can use this
                _status: 404,

                // tell autoresponse that the json data will be processed by smarty processor
                _process: "smarty",

                // the smarty template name will be rendered
                _tpl: "a/b.tpl",

                // define the template data to be applied to smarty template file
                _data: {
                    extData: [],
                    tplData: {}
                }
            };
        };
        ```

    -   output json with smarty render result


        ```javascript
        module.exports = function (path, queryParam, postParam) {
            return {
                // the json data will be processed by smarty processor
                _process: 'smarty',

                filters: [],

                // the smarty render result will be replaced as the value of `filterTpl`
                filterTpl: {
                     // the smarty template name will be rendered
                     _tpl: 'filter.tpl',
                     // define the template data to be applied to smarty template file
                     _data: {
                         extData: [],
                         tplData: {}
                     }
                }
            };
        };
        ```

## Using mock helper method <a name="helper"></a>

By default, if you use js file to mock, you can access `mock` global variable in your mock file.

The following methods are provided by default:

-   `mock._`: [lodash](https://lodash.com/docs) variable

-   `mock.m`: [dayjs](https://github.com/iamkun/dayjs) variable，using [moment](http://momentjs.com/docs/) before the `0.3.0` version

-   `mock.fake(format, locale)`: the encapsulation of [faker](https://github.com/Marak/faker.js/)

    ```javascript
    // more api and variable name, please refer faker api docs
    mock.fake("{{name.firstName}}-{{name.lastName}}");
    ```

-   `mock.fakeCN(format)`: generate chinese locale random information

-   `mock.fakeEN(format)`: is equivalent to `mock.fake(format)`, generate english locale random information

-   `mock.faker(locale)`: get `faker` instance with the specified locale, the locale argument is default english

More details, please refer to the [autoresponse-config.js](https://github.com/wuhy/autoresponse/blob/master/lib/autoresponse-config.js).


history.md
0.4.1 / 2020-01-21
===================

* [^] 修复 `logLevel` 选项设置无效问题


0.4.0 / 2020-01-21
===================

* [^] 更新依赖，要求 `Node >= 8`
* [^] mock 默认辅助的工具方法，使用更轻量级的日期库 [dayjs](https://github.com/iamkun/dayjs) 替代 [moment](http://momentjs.com/docs/) 日期库


0.3.0 / 2018-11-30
===================

* [^] 更新依赖，要求 `Node >= 6`


0.2.3 / 2018-11-30
===================

* [^] 修复请求路径重写出现 `null` 问题 [8949615](https://github.com/wuhy/autoresponse/commit/8949615)
* [^] 修复查询参数正则 [5c7d92c](https://github.com/wuhy/autoresponse/commit/5c7d92c)

0.1.3 / 2015-08-19
===================

  * [^] mockDataTpl路径之前相对于配置文件路径，现在改成相对于responseDir
  * [+] 增加默认的mock数据模板：json 和 js 类型
  * [^] autoMock默认设为true
  * [+] 配置项支持直接传入match的路径、正则或function(reqPath, context){return {match: '', mock: ''}}
  * [+] timeout全局配置包括mock配置的timeout支持随机功能，语法类似于：timeout: '0, 100'
  * [+] 增加特殊响应数据字段：

    - _timeout：（同timeout）

    - _status: 增加设置响应的http状态码

    - _header: 增加自定义的响应数据头信息

    - _data：增加该数据字段，可以把所有要响应的数据统一放在该字段下

    - _process: 增加当前 json 数据需要预处理的后处理器类型，比如 `smarty`

    - _tplPath: 要渲染的模板路径，比如 `smarty` 的模板路径

  * [^] autoresponse-config.js配置文件变化监听用 `chokidar` 模块实现，替换掉不稳定的 `fs.watch` 接口
  * [+] 增加 php 预处理器
  * [+] 增加 smarty 预处理器：可以通过定义 json 数据来直接输出 smarty 模板，也支持部分数据项值替换为 smarty 模板输出
  * [^] 调整下默认的 js 预处理器的 mock 数据方法的 查询参数 和 post 参数顺序
  * [+] 增加 mock 一些辅助的工具方法，通过autoresponse-config.js如下配置选项定义：

    ```
     helperName: 'mock' // 指定 mock 助手工具方法要挂载的全局变量名称，默认值 `mock`
     helper: {_: 'lodash', m: 'moment'} // 这样可以访问：`mock._.merge()`
    ```

    扩展助手工具方法，除了通过该选项配置，也可以通过 `mock.inject({a: function() {}})` 方式添加
    此外还封装了 [faker](http://marak.com/faker.js/) 库来提供随机数据生成方法




examples/start.js
var http = require('http');
var connect = require('connect');
var app = connect();

// watch: 监控配置文件变化是否重新reload
// logLevel: 设置控制台打印 log level，如果不想输出太多log信息，可以把log level调高，建议 error
// refer: https://github.com/nomiddlename/log4js-node
var autoresponse = require('../index')({
    watch: true, logLevel: 'debug', handlers: [
        // 增加自定义处理器
        function (context, options, next) {
            console.log('custom handle:', context.content);
            next();
        }
    ]
});
app.use(autoresponse);
app.use('/account/getUserInfo', function (req, res) {
    res.writeHead(200, {
        'content-type': 'application/json;charset=UTF-8'
    });

    res.end(JSON.stringify({
        name: 'Jack',
        age: 33
    }));
});

app.use('/mock/getUserInfo', function (req, res) {
    res.writeHead(200, {
        'content-type': 'application/json;charset=UTF-8'
    });

    res.end(JSON.stringify({
        name: 'I\'m not Jack, 0_0',
        age: 44
    }));
});

app.use('/data/test.php', function (req, res) {
    res.writeHead(200, {
        'content-type': 'text/html;charset=UTF-8'
    });

    res.end('<html><head><title>Test PHP Proxy</title></head><body>Hello PHP</body></html>');
});

var serveStatic = require('serve-static');
app.use(serveStatic('./fixtures'));

var httpServer = http.createServer(app);
var port = 9090;
httpServer.listen(port, function () {
    console.log('Http server started on port: %d ...', port);
});
